### ISSUE TITLE
How do we support synchronous wait?

### ISSUE THREAD

**Issue by kelunik** [async-interop/event-loop#47]
Wait could be implemented like this:
https://gist.github.com/kelunik/4163c84839ce374a173e57926a608b58

Issue is that we would either have to make `when` throw under certain conditions or swallow the exception until the loop is continued, e.g. by the next `wait`, see http://chat.stackoverflow.com/transcript/message/30615565#30615565

**Comment by bwoebi** [comment]
Yes, we should. There are always some ways you can shoot yourself into your own feet. That should not prevent you from making a nice feature generally available.

The only thing necessary is having well-defined behavior for that (and any other edge-) case, which is being discussed in async-interop/awaitable#7.

**Comment by kelunik** [comment]
I think it's something we can discuss there.

**Comment by kelunik** [comment]
Support for `wait` breaks our nice `Loop::execute` scoping. I'm not really sure if we really want to support `wait`. People wanting to mix things could still just put their entire application inside the loop, e.g. by using `Loop::execute(function () { require __DIR__ . "/app.php"; });`. The disadvantage is, that they cannot just use some async API and use `wait`, they would probably have to restructure it completely up to the root to support the async waiting.

**Comment by bwoebi** [comment]
If we put Loop::execute() around, we basically have won nothing particular. It's still a scope around the whole program in this case.
That's pretty arbitrary and nonsense, _IMO_.

**Comment by kelunik** [comment]
The difference is that it's around the program. It's not global as soon as you require some event loop implementation. But it probably doesn't make much sense for mixing sync and async, because you have no possibility to wait for a specific thing.

**Comment by bwoebi** [comment]
It's not actually global, but it's defacto global as it encompasses the whole program. (IMO)

**Comment by joelwurtz** [comment]
IMO a wait method does not make sense and maybe the problem can be reversed by thinking about the use case of a wait : 

Generally we need to do a wait because we want to use a result of a asynchronous call in another asynchronous call so the thing that is really needed is a dependency system between call (like thread synchronization) is the wait method the only way to do that ?

**Comment by WyriHaximus** [comment]
The wait method is to call async code in sync code and wait for the result using a nice syntax. For waiting on something async in async code we have promises/awaitables/observables.

A wait method has several uses cases, included but not limited to: testing promises in unit tests, waiting for a bunch of async calls in a sync request. The reactphp environment has [`clue/block-react`](https://github.com/clue/php-block-react#await) which makes this possible and honestly it is a really nice convenience method to have.

**Comment by kelunik** [comment]
Just a note: We do not need wait for Tests, just for async in sync context.

**Comment by WyriHaximus** [comment]
> Just a note: We do not need wait for Tests, just for async in sync context

Only mentioned it as simple example ☺️

**Comment by bwoebi** [comment]
Can we please close this now?

At least the way the code is designed now, it's easily supported. Or is there really much benefit at explicitly forbidding it at the cost of it being hard to use in sync contexts?

**Comment by WyriHaximus** [comment]
Closing this, please reopen if anyone feels the need to discuss it more

**Comment by kelunik** [comment]
I'd like to remove `Loop::get`, people should use `Loop::execute`. https://github.com/ReactiveX/RxPHP/issues/101 suggests to use `Loop::get()->run()`, I've seen others using `Loop::get` to inject the loop as dependency. Something like that probably results in lazy injection somewhere leading to multiple loops, exactly the thing we want to prevent.

I get that we need a way to continue with the same loop for multiple `wait`s. People wanting to use `wait` should either always pass a loop to it or the implementation may use black magic to access the current driver and continue with it.

```php
$class = new ReflectionClass(Loop::class);
$driver = $class->getStaticProperties()["driver"];
```

**Comment by bwoebi** [comment]
I feel like this is rather a matter of education. [Perhaps add a comment to the docblock?]

Also, for what RxPHP is proposing it, it isn't bad by default, it feels like they are proposing it to use with wait() like semantics inside sync code. I personally think they should hide it in a function, but that we can debate about.

Also, officially proposing to use black magic makes it defacto-API. It changes nothing except that it is a few bytes more to write.

**Comment by kelunik** [comment]
> I feel like this is rather a matter of education.

We said that often in PHP, and how often did it really work?

> Also, officially proposing to use black magic makes it defacto-API. It changes nothing except that it is a few bytes more to write.

It does change something. People probably won't use it to share the loop in their DIC.

`wait` seems to be the only reason why this method exists.

**Comment by bwoebi** [comment]
And getting the underlying Loop, a function which we didn't want to expose on the Loop static…

**Comment by kelunik** [comment]
In that case I'd add `getHandle` to the accessor. Maybe we should do that anyway. Consistency is more important than a smaller autocomplete list.

**Comment by davidwdan** [comment]
@kelunik as far as RxPHP v2 is concerned, I've removed the example with `Loop::get()->run()`.

**Comment by bwoebi** [comment]
@kelunik I find it incredibly terrible to force the use of black magic here … Sorry, I don't feel like dropping Loop::get() is a solution for me…

**Comment by joshdifabio** [comment]
Yeah, I have to agree that forcing people to use reflection would not be nice. On a side note, why isn't the method called `getDriver()`?

**Comment by bwoebi** [comment]
I'm all in favor of having a more specific name, if it clears up what's happening.

But we have setFactory, execute, stop, get etc.… and not setDriverFactory, executeDriver (or even runDriver), stopDriver etc.
Including `Driver` in the getter name is just purely redundant and inconsistent with the other method names.

**Comment by joshdifabio** [comment]
I agree that perhaps it would be inconsistent to have `getDriver` and `setFactory`, but I think it would be reasonable to have `getDriver` and `setDriverFactory` without renaming the other methods, as I think their names actually make sense.

In English, they read to me as:
- `Loop::execute($callback)` - execute the callback
- `Loop::stop()` - stop the loop
- `Loop::defer($callback)` - defer the callback
- `Loop::get()` - get the loop. But this is obviously not what the method does.
... etc

It's only really `get()` and, to a lesser extent, `setFactory()` which I find misleading. I don't feel strongly though, and I'm going off topic, sorry!

**Comment by bwoebi** [comment]
> Loop::get() - get the loop. But this is obviously not what the method does.

Well, what is _the_ loop? IMHO the driver **is** _the_ loop.

**Comment by davidwdan** [comment]
It looks like multiple things are being discussed on this issue:
1. Synchronous wait support
2. Remove `Loop::get` and use reflection to access the driver
3. Rename `Loop::get` to `Loop::getFactory`

Which of these items are blocking a v1 release?

**Comment by kelunik** [comment]
> 3. Rename `Loop::get` to `Loop::getFactory`

`Loop::get` to `Loop::getDriver`.

> Synchronous wait support
> Remove `Loop::get` and use reflection to access the driver

That's one point, we need either `Loop::get` or reflection to support wait with reused loops.

I think both should be resolved, but they're mostly a simple vote to resolve.

**Comment by kelunik** [comment]
@joshdifabio If you want to discuss renaming `Loop::get`, please open another issue.

**Comment by bwoebi** [comment]
@joshdifabio If you do not feel very strongly about this, I'd love to avoid yet another discussion...

**Comment by joshdifabio** [comment]
I think Bob's explanation of the current names is reasonable; I don't feel a need to discuss it further.

**Comment by kelunik** [comment]
I think we're leaving this as is, wait seems to be important enough for now.
